### DISCUSSION TITLE
Can We Use App.Storage To Store Table / Aggrid Object In Session Data?

### DISCUSSION THREAD

**Discussion by clueple** [zauberzeug/nicegui#987] [Q&A] [answered]
### Question

I'd like to know how to store the session data (e.g. see my codes below and the image for details) of a table (e.g. ui.aggrid()) so I can export to a csv in the local downloads folder.

I'm pretty sure the .call_api_method('exportDataAsCsv') works as I tested changing the table to link to an absolute path of my uploaded csv (not shown in the codes below)

```py
from nicegui import ui, events
import pandas as pd 
from io import BytesIO

df = pd.DataFrame()

def handle_upload(e:events.UploadEventArguments):
    # m_name = e.name
    m_content = e.content
    # m_type = e.type
    m_content_str = e.content.read() # text string
    b_content = BytesIO(m_content_str)

    read_data = pd.read_csv(b_content)

    global df
    df = pd.concat([df, read_data], ignore_index=True)

    table.from_pandas(df).classes('max-h-40')


ui.upload(on_upload = handle_upload, auto_upload=True)


table = ui.aggrid({'rowData': [],'columnDefs':[]}).style('display:none')
# table = ui.aggrid({'rowData': [],'columnDefs':[]})
table.update()

ui.button('export', on_click = lambda: table.call_api_method('exportDataAsCsv'))

ui.run()
```
When running the above code, I get a blank csv (when clicking the 'export' button)

![Capture](https://github.com/zauberzeug/nicegui/assets/19367464/9d375b5b-a2c6-4539-990b-2331b09b3ace)

As you can see from the image, the 'table' (an ui.aggrid()) object can be in 2 different states:

b1) before upload, it should be blank, and I set it to display:none 
b2) after upload, it should show the result of the click event (see code, function name: 'handle_upload')

Seems like it's loading state (b1) instead of (b2).

That's why I'd like to know if there's any way to store the session data for the df and the table.

Many thanks for your help.

**Comment by falkoschindler** [comment, marked as answer]
Hi @clueple! Here are my thoughts about your code:

1. I'm not sure if this is the right way to load uploaded CSV data into pandas:
    ```py
    m_content_str = e.content.read() # text string
    b_content = BytesIO(m_content_str)
    read_data = pd.read_csv(b_content)
    ```
    I'd suggest to start here and verify that the data is correctly loaded into the dataframe.
2. `from_pandas` creates a _new_ UI element. It's supposed to be used something like this:
    ```py
    global table
    table = ui.aggrid.from_pandas(df).classes('max-h-40')
    ```
3. You seem to be using the AG Grid element only for an easy way to export the combined data. I think it would be cleaner and easier to write the new dataframe to a CSV file and trigger the download using `ui.download`.

**Reply by clueple** [reply]
Thanks, @falkoschindler, for your suggestion.

Haven't tested myself, I've seen tutorials regarding (table / aggrid update on nicegui) that does the following (just the pattern of coding):

```py
# create a blank data table object to store the original data, change of data, 'df'
df = [ ]

# create the blank table object to display 'the latest state of the 'df' object':
table = [ ]

# create fields input area, the tutorial author uses ui.notify () , for the popup of fields input;  for simplicity, I just use a fixed pre-defined text_input fields without any styling

field1_input = ui.input('field1')
field2_input = ui.input('field2')
fieldn_input = ui.input('fieldn')

""" create upload function and the upload area for csv upload  """
async def handle_upload(e: events.UploadEventsArguments):
   read_data = pd.read_csv( BytesIO(e.content.read())  )
   return read_data
   table.append( pd.concat( [df, read_data] )  )

upload_df = ui.upload(on_upload= handle_upload, auto_upload=True )

# create function to add data and the button to add data
async def add_data():
   add_data = { ''field1': field1_input.value , 'field2': field2_input.value, 'fieldn': filedn_input.value   } 
   df= pd.append(df, add_data )
   table.append(df)

add_btn = ui.button('add', on_click= add_data )

# create function to delete data and the button to delete data
async def del_data():
     # create a selection of row index of the 'df' table
     # when the row index is selected, delete the row from the 'df' 
     # read the 'df' in the 'table' 

# create function to edit data and the button to edit data
async def edit_data():
     # create a selection of row index of the 'df' table
     # when the row index is selected, pop up the buttons to input the row of 'df' to be edited
     # read the 'df' in the 'table' 

edit_btn = ui.button('delete', on_click= edit_data )

#export whatever is displayed on 'table' to a csv file in the download folder in the local machine
ui.button('export to csv', on_click = lambda: table.call_api_button('exportDataAsCsv') )

ui.run()
```

Hopefully, I can share my successful result here.

Thanks for your tips.
